pydantic_ai_slim/pydantic_ai/_agent_graph.py

```diff
@@ -864,6 +863,24 @@ async def process_tool_calls(  # noqa: C901
         output_final_result.append(final_result)
 
 
+def _check_tool_call_limits(
+    tool_manager: ToolManager[DepsT],
+    tool_calls: list[_messages.ToolCallPart],
+    usage_limits: _usage.UsageLimits | None,
+) -> None:
+    """Check if executing the tool calls would exceed the limit."""
+    if usage_limits is None or usage_limits.tool_calls_limit is None:
+        return
+
+    current_tool_calls = tool_manager.ctx.usage.tool_calls if tool_manager.ctx is not None else 0
+    projected_tool_calls = current_tool_calls + len(tool_calls)
+
+    if projected_tool_calls > usage_limits.tool_calls_limit:
+        projected_usage = deepcopy(tool_manager.ctx.usage) if tool_manager.ctx else _usage.RunUsage()
+        projected_usage.tool_calls = projected_tool_calls
```

Comment:
If we pass `usage` in, we don't need the check above anymore, and can do it like this:

```py
projected_usage = deepcopy(usage)
projected_usage.incr(RequestUsage(tool_calls=len(tool_calls))
```

That's then also likely easy enough that it can be inlined instead of being in a separate function